syntax = "proto3";

package distributedstore;

service KeyValueStore {
  rpc put(PutRequest) returns (PutResponse);
  rpc get(GetRequest) returns (GetResponse);

  rpc slowDown(SlowDownRequest) returns (SlowDownResponse);
  rpc restore(RestoreRequest) returns (RestoreResponse);

  rpc registerSlave(RegisterSlaveRequest) returns (RegisterSlaveResponse);
  rpc registerNode(RegisterNodeRequest) returns (RegisterNodeResponse);

  rpc askVote(AskVoteRequest) returns (AskVoteResponse);
  rpc canCommit(CommitRequest) returns (CommitResponse);
  rpc doCommit(CommitRequest) returns (Empty);
}

message PutRequest {
  string key = 1;
  string value = 2;
}

message PutResponse {
  bool success = 1;
}

message GetRequest {
  string key = 1;
}

message GetResponse {
  string value = 1;
  bool found = 2;
}

message SlowDownRequest {
  int32 seconds = 1;
}

message SlowDownResponse {
  bool success = 1;
}

message RestoreRequest {}

message RestoreResponse {
  bool success = 1;
}

message RegisterSlaveRequest {
  string address = 1;
}

message RegisterSlaveResponse {
  bool success = 1;
  map<string, string> state = 2;
}

message RegisterNodeRequest {
  string address = 1;
}

message RegisterNodeResponse {
  bool success = 1;
  map<string, string> state = 2;
}

message AskVoteRequest {
  string key = 1;
}

message AskVoteResponse {
  int32 weight = 1;
}

message CommitRequest {
  string key = 1;
  string value = 2;
}

message CommitResponse {
  bool can_commit = 1;
}

message Empty {}